### PR TITLE
Internal: Adjust experiments to support better dependency between feature [ED-11810]

### PR DESCRIPTION
**Pull request by Nevoss**

### CODE REVIEW

File: tests/phpunit/elementor/core/utils/test-collection.php [hunk 1]
```diff
@@ -469,4 +469,39 @@ public function test_each__breaks_on_false() {
...
+
+	public function test_find__with_value() {
+		// Arrange.
+		$collection = Collection::make( [ 'a1', 'b2', 'c3', 'd4' ] );
+
+		// Act.
+		$result = $collection->find( 'b2' );
+
+		// Expect.
+		$this->assertEquals( 'b2', $result );
```
Comment (StyleShit) [inline tests/phpunit/elementor/core/utils/test-collection.php:494]: hmm, something is weird
shouldn't "find" return a boolean?
I mean... you're looking for a value that you already have
maybe we should have "find" for callback and "contains" for values?

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: core/experiments/manager.php [hunk 2]
```diff
@@ -386,7 +388,7 @@ private function add_default_features() {
 				Sections, Inner Sections and Columns and be able to edit them. Ready to give it a try? Check out the %3$sFlexbox playground%4$s.',
 				'elementor'
 			), '<a target="_blank" href="https://go.elementor.com/wp-dash-flex-container/">', '</a>', '<a target="_blank" href="https://go.elementor.com/wp-dash-flex-container-playground/">', '</a>'),
-			'release_status' => self::RELEASE_STATUS_RC,
+			'release_status' => self::RELEASE_STATUS_STABLE,
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/experiments/manager.php:391]: Done

File: core/experiments/manager.php [hunk 3]
```diff
@@ -882,6 +884,70 @@ private function create_dependency_class( $dependency_name, $dependency_args ) {
...
+			}
+
+			$feature = $this->get_features( $item );
+
+			if ( ! $feature ) {
+				return;
+			}
+
+			foreach ( $feature['dependencies'] ?? [] as $dep ) {
+				// Check if exists in allowed options
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/experiments/manager.php:918]: Done

File: core/experiments/manager.php [hunk 4]
```diff
@@ -882,6 +884,70 @@ private function create_dependency_class( $dependency_name, $dependency_args ) {
...
+	 */
+	private function sort_allowed_options_by_dependencies( $allowed_options ) {
+		if ( ! isset( $allowed_options['elementor'] ) ) {
+			return $allowed_options;
+		}
+
+		$sorted = Collection::make();
+
+		$sort = function ( $item ) use ( &$sorted, &$sort ) {
+			if ( $sorted->find( $item ) ) {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/experiments/manager.php:907]: Changed to `contains
`

File: core/experiments/manager.php [hunk 5]
```diff
@@ -882,6 +884,70 @@ private function create_dependency_class( $dependency_name, $dependency_args ) {
...
+				$name = is_string( $dep ) ? $dep : $dep->get_name();
+
+				$sort( $name );
+			}
+
+			$sorted->push( $item );
+		};
+
+		foreach ( $allowed_options['elementor'] as $option ) {
+			if ( 0 !== strpos( $option, static::OPTION_PREFIX ) ) {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply core/experiments/manager.php:928]: Done

File: tests/phpunit/elementor/core/experiments/test-manager.php [hunk 6]
```diff
@@ -566,6 +566,72 @@ public function test_validate_dependency__deactivates_an_experiment_when_its_dep
...
+				'elementor' => [
+					'option_b',
+					'elementor_experiment-test_B',
+					'elementor_experiment-test_D',
+					'elementor_experiment-test_E',
+					'option_c',
+					'elementor_experiment-test_A',
+				],
+			]
+		);
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply tests/phpunit/elementor/core/experiments/test-manager.php:608]: DONE

File: tests/phpunit/elementor/core/experiments/test-manager.php [hunk 7]
```diff
@@ -566,6 +566,72 @@ public function test_validate_dependency__deactivates_an_experiment_when_its_dep
...
+		// Arrange.
+		$this->experiments->add_feature( [
+			'name' => 'test_A',
+		] );
+
+		$this->experiments->add_feature( [
+			'name' => 'test_B',
+			'dependencies' => [
+				'test_A',
+				'test_C',
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply tests/phpunit/elementor/core/experiments/test-manager.php:579]: done

File: tests/phpunit/elementor/core/experiments/test-manager.php [hunk 8]
```diff
@@ -566,6 +566,72 @@ public function test_validate_dependency__deactivates_an_experiment_when_its_dep
...
+			],
+		] );
+
+		$this->experiments->add_feature( [
+			'name' => 'test_D',
+			'dependencies' => [ 'test_A' ],
+		] );
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply tests/phpunit/elementor/core/experiments/test-manager.php:586]: Done

File: tests/phpunit/elementor/core/utils/test-collection.php [hunk 9]
```diff
@@ -469,4 +469,39 @@ public function test_each__breaks_on_false() {
 		// Act
 		$collection->each( [ $mock, 'each_callback' ] );
 	}
+
+	public function test_find() {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply tests/phpunit/elementor/core/utils/test-collection.php:473]: Changed

File: tests/phpunit/elementor/core/utils/test-collection.php (hunk 1 above)
Comment (Nevoss) [reply tests/phpunit/elementor/core/utils/test-collection.php:494]: Done

File: tests/phpunit/elementor/core/experiments/test-manager.php [hunk 10]
```diff
@@ -566,6 +566,127 @@ public function test_validate_dependency__deactivates_an_experiment_when_its_dep
...
+		$this->assertEqualSets(
+			[
+				'elementor' => [
+					'elementor_experiment-test_C',
+					'elementor_experiment-test_A',
+					'elementor_experiment-test_B',
+				],
+			],
+			$result
+		);
```
Comment (StyleShit) [inline tests/phpunit/elementor/core/experiments/test-manager.php:682]: hmm
do we need to fail for circular deps? what's the order logic in this case?

File: tests/phpunit/elementor/core/utils/test-collection.php [hunk 11]
```diff
@@ -469,4 +469,36 @@ public function test_each__breaks_on_false() {
...
+		// Act.
+		$result1 = $collection->contains( 'b2' );
+
+		$result2 = $collection->contains( function ( $item ) {
+			return 'd4' === $item;
+		} );
+
+		$result3 = $collection->contains( 'b3' );
+
+		// Expect.
```
Comment (StyleShit) [inline tests/phpunit/elementor/core/utils/test-collection.php:499]:
```suggestion
		// Assert.
```

**Review by StyleShit — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/core/experiments/test-manager.php [hunk 12]
```diff
@@ -566,6 +566,127 @@ public function test_validate_dependency__deactivates_an_experiment_when_its_dep
...
+
+		// Teardown.
+		$this->experiments->remove_feature( 'test_A' );
+		$this->experiments->remove_feature( 'test_B' );
+		$this->experiments->remove_feature( 'test_C' );
+		$this->experiments->remove_feature( 'test_D' );
+		$this->experiments->remove_feature( 'test_E' );
+	}
+
+	public function test_test_sort_allowed_options_by_dependencies__recursive_deps() {
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply tests/phpunit/elementor/core/experiments/test-manager.php:643]: Done

File: tests/phpunit/elementor/core/experiments/test-manager.php (hunk 10 above)
Comment (Nevoss) [reply tests/phpunit/elementor/core/experiments/test-manager.php:682]: ?

File: tests/phpunit/elementor/core/utils/test-collection.php [hunk 13]
```diff
@@ -469,4 +469,36 @@ public function test_each__breaks_on_false() {
...
+	public function test_find() {
+		// Arrange.
+		$collection = Collection::make( [ 'a1', 'b2', 'c3', 'd4' ] );
+
+		// Act.
+		$result = $collection->find( function( $item ) {
+			return strpos( $item, '3' ) !== false;
+		} );
+
+		// Expect.
```
[the remark this answers is not in the document]
Comment (Nevoss) [reply tests/phpunit/elementor/core/utils/test-collection.php:482]: Done

File: tests/phpunit/elementor/core/utils/test-collection.php (hunk 11 above)
Comment (Nevoss) [reply tests/phpunit/elementor/core/utils/test-collection.php:499]: Done

**Review by StyleShit — APPROVED** [review]

**Pull request merged**